src/core/query.ts

```diff
@@ -0,0 +1,653 @@
+import {
+  isServer,
+  functionalUpdate,
+  cancelledError,
+  isDocumentVisible,
+  noop,
+  Console,
+  getStatusBools,
+  shallowEqual,
+  Updater,
+} from './utils'
+import { QueryInstance } from './queryInstance'
+import {
+  ArrayQueryKey,
+  QueryStatus,
+  InfiniteQueryConfig,
+  QueryFunction,
+  InitialDataFunction,
+  QueryConfig,
+} from './types'
+import { QueryCache } from './queryCache'
+
+// TYPES
+
+interface QueryInitConfig<TResult, TError> {
+  queryCache: QueryCache
+  queryKey: ArrayQueryKey
+  queryHash: string
+  config: QueryConfig<TResult, TError>
+  notifyGlobalListeners: (query: Query<TResult, TError>) => void
+}
+
+export interface QueryState<TResult, TError> {
+  canFetchMore?: boolean
+  data?: TResult
+  error: TError | null
+  failureCount: number
+  isError: boolean
+  isFetching: boolean
+  isFetchingMore?: false | 'previous' | 'next'
```

Comment:
If we are using this type at multiple places (I think in Infinite query), let's have them as explicit type in core types.

Reply:
Yes good one, created a type for it